File: ui/src/events.ts
/**
 * The server's side of the conversation, for things that are not answers.
 *
 * The panel does not poll for changes; it is told. Every section names the
 * topic it displays and is re-rendered when that topic fires, which makes
 * being current the default rather than something each panel has to remember
 * to arrange. A section that forgets to poll is stale forever; a section that
 * forgets to name a topic simply does not refresh, which is visible the first
 * time anyone looks at it.
 *
 * EventSource rather than a WebSocket: nothing travels the other way, and it
 * reconnects by itself after a server restart or a sleeping laptop, which is
 * most of what a hand-written socket client ends up being.
 */

export type Topic =
  | 'notes'
  | 'timers'
  | 'memory'
  | 'conversation'
  | 'settings'
  | 'client'
  | 'turn'
  | 'ops';

export interface ServerEvent {
  topic: Topic;
  /** Only on a timer that has come due. */
  rang?: string;
  at?: number;
  /** Only on 'client': what the tray's push-to-talk is doing. */
  state?: string;
  /** Only on 'client': the tray asking its window to come forward. */
  show?: boolean;
  /** Only on 'client': a stop was asked for, so any playback should end. */
  interrupt?: boolean;
  /** Only on 'turn': which exchange, and what just happened in it. */
  id?: number;
  phase?: 'heard' | 'sentence' | 'done';
  /** On 'heard', what was said to her; on 'sentence', what she said. */
  text?: string;
  via?: 'voice' | 'text';
  interrupted?: boolean;
}

interface Handlers {
  onEvent: (event: ServerEvent) => void;
  /** Connected or not, for the status line. */
  onConnection?: (live: boolean) => void;
}

export function listen(handlers: Handlers): () => void {
  let source: EventSource | null = null;
  let stopped = false;

  const open = (): void => {
    if (stopped) return;
    source = new EventSource('/events');

    source.addEventListener('open', () => {
      handlers.onConnection?.(true);
    });

    source.addEventListener('error', () => {
      // Not fatal and not worth reporting loudly: EventSource is already
      // retrying, and a server restart takes this path every time.
      handlers.onConnection?.(false);
    });

    source.addEventListener('message', (event: MessageEvent<string>) => {
      let parsed: ServerEvent;
      try {
        // oxlint-disable-next-line typescript/no-unsafe-type-assertion
        parsed = JSON.parse(event.data) as ServerEvent;
      } catch {
        return;
      }
      if (typeof parsed.topic === 'string') handlers.onEvent(parsed);
    });
  };

  // Opened once the page has loaded rather than during module evaluation.
  //
  // Only a small thing: browsers allow six connections per origin over
  // HTTP/1.1, and this one is held for the life of the tab, so there is no
  // reason for it to compete with the module graph and the stylesheet while
  // those are still arriving. It does not block the load event — that was
  // measured, after a page that would not load turned out to have nothing to
  // do with this and everything to do with WSL2 forwarding.
  if (document.readyState === 'complete') open();
  else globalThis.addEventListener('load', open, { once: true });

  return () => {
    stopped = true;
    source?.close();
  };
}

File: server/events.py
"""Server to client push, over Server-Sent Events.

Every change the panel might be showing is announced here, and the panel
re-reads what it needs. That is the default rather than something added per
feature: a panel that polls only what someone remembered to make it poll is a
panel where the next thing added is stale, which is how both the timer list
and the notes list ended up frozen on whatever was true when they opened.

Events carry a topic and almost never any data. Telling a client what changed
and letting it ask is far less fragile than shipping state down two paths and
hoping they agree — the client already has an endpoint that returns the truth.

Everything else here is request-shaped: the client asks and the server
answers. A timer is the first thing that has to travel the other way — it
comes due whether or not anyone is asking — so it needs a channel that is
already open when it fires.

SSE rather than a WebSocket, because nothing ever travels client to server on
it. EventSource reconnects on its own where a socket needs hand-written
backoff, it is plain HTTP so it can be read with curl and needs no protocol
upgrade, and it costs no dependency — uvicorn cannot serve a WebSocket at all
without `websockets` or `wsproto` installed. The things a socket would buy
here (binary frames, client-to-server traffic, more than six connections per
origin) are all things this does not do.

The hub is deliberately forgetful. It holds no history: a client that is not
connected when a timer goes off is not told later, because a kitchen timer
that rings ten minutes after the pasta is worse than one that never rang.
"""

import asyncio
import json
import logging
from collections.abc import AsyncIterator

log = logging.getLogger("naka.events")

# Long enough to stay quiet, short enough that a browser or any proxy in
# between does not decide the connection died.
KEEPALIVE_SECONDS = 20

# Per client. A panel this far behind is not reading, and unbounded queues
# would grow without limit behind a tab the machine has suspended.
BACKLOG = 32


class Hub:
    def __init__(self) -> None:
        self._clients: set[asyncio.Queue] = set()

    @property
    def listeners(self) -> int:
        return len(self._clients)

    def subscribe(self) -> asyncio.Queue:
        queue: asyncio.Queue = asyncio.Queue(maxsize=BACKLOG)
        self._clients.add(queue)
        log.info("client connected (%d listening)", len(self._clients))
        return queue

    def unsubscribe(self, queue: asyncio.Queue) -> None:
        self._clients.discard(queue)
        log.info("client gone (%d listening)", len(self._clients))

    def dispatch(self, event: dict) -> None:
        """Fan out. Called on the loop thread; never blocks."""
        for queue in list(self._clients):
            try:
                queue.put_nowait(event)
            except asyncio.QueueFull:
                # Dropped rather than awaited: one unresponsive tab must not
                # hold up the ring for every other panel that is open.
                log.warning("dropping event for a client that is not reading")


hub = Hub()

_loop: asyncio.AbstractEventLoop | None = None


def bind(loop: asyncio.AbstractEventLoop) -> None:
    """Remember the loop, so publish() can be called from anywhere."""
    global _loop
    _loop = loop


def publish(topic: str, **data) -> None:
    """Announce a change. Safe from any thread.

    Tools run in a worker thread via asyncio.to_thread, so the obvious direct
    call would be touching loop state from off the loop — which fails rarely
    and confusingly rather than immediately.
    """
    event = {"topic": topic, **data}
    if _loop is None:
        return
    try:
        _loop.call_soon_threadsafe(hub.dispatch, event)
    except RuntimeError:
        # The loop is closing, which happens during shutdown. Nothing is
        # listening by then anyway.
        pass


def frame(event: dict) -> str:
    return f"data: {json.dumps(event)}\n\n"


async def stream() -> AsyncIterator[str]:
    """One client's event stream, for a text/event-stream response."""
    queue = hub.subscribe()
    try:
        # Sent immediately so the response headers leave now rather than when
        # the first timer happens to fire, which is what makes EventSource
        # consider the connection open.
        #
        # The retry hint shortens the window this cannot cover. Nothing is
        # replayed to a client that reconnects — a kitchen timer ringing ten
        # minutes late is worse than one that never rang — so a change made
        # while a client is away is a change it never hears about. The client
        # re-reads on every open for exactly that reason; this just makes the
        # gap it has to repair a short one. Browsers default to about three
        # seconds and back off from there.
        yield "retry: 2000\n\n"
        yield ": connected\n\n"
        while True:
            try:
                event = await asyncio.wait_for(queue.get(),
                                               timeout=KEEPALIVE_SECONDS)
            except TimeoutError:
                yield ": keepalive\n\n"
                continue
            yield frame(event)
    finally:
        # Reached when the client disconnects and the generator is closed.
        hub.unsubscribe(queue)
